File: app/src/main/java/net/squanchy/favorites/FavoritesPageView.java
package net.squanchy.favorites;

import android.app.Activity;
import android.content.Context;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.widget.Toolbar;
import android.util.AttributeSet;
import android.view.View;

import net.squanchy.R;
import net.squanchy.analytics.Analytics;
import net.squanchy.analytics.ContentType;
import net.squanchy.favorites.view.FavoritesListView;
import net.squanchy.home.HomeActivity;
import net.squanchy.home.Loadable;
import net.squanchy.navigation.Navigator;
import net.squanchy.schedule.ScheduleService;
import net.squanchy.schedule.domain.view.Event;
import net.squanchy.schedule.domain.view.Schedule;
import net.squanchy.schedule.view.ScheduleViewPagerAdapter;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;

import static net.squanchy.support.ContextUnwrapper.unwrapToActivityContext;

public class FavoritesPageView extends CoordinatorLayout implements Loadable {

    private View progressBar;
    private Disposable subscription;
    private ScheduleService service;
    private Navigator navigate;
    private Analytics analytics;
    private FavoritesListView favoritesListView;
    private View emptyViewSignedIn;
    private View emptyViewSignedOut;

    public FavoritesPageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public FavoritesPageView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        progressBar = findViewById(R.id.progressbar);
        favoritesListView = (FavoritesListView) findViewById(R.id.favorites_list);
        emptyViewSignedIn = findViewById(R.id.empty_view_signed_in);
        emptyViewSignedOut = findViewById(R.id.empty_view_signed_out);
        emptyViewSignedOut.setOnClickListener(view -> sequestSignIn());

        setupToolbar();

        if (!isInEditMode()) {
            Activity activity = unwrapToActivityContext(getContext());
            FavoritesComponent component = FavoritesInjector.obtain(activity);
            service = component.service();
            navigate = component.navigator();
            analytics = component.analytics();
        }
    }

    private void sequestSignIn() {
        // ⚠️ HACK this is DIRTY and HORRIBLE but it's the only way we can ship this
        // without rewriting the whole data layer. Sorry. I swear, we know it sucks
        // and we want to fix this ASAP.
        HomeActivity activity = (HomeActivity) unwrapToActivityContext(getContext());
        activity.requestSignIn();
    }

    private void setupToolbar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.activity_favorites);
        toolbar.inflateMenu(R.menu.homepage);
        toolbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_search:
                    navigate.toSearch();
                    return true;
                case R.id.action_settings:
                    navigate.toSettings();
                    return true;
                default:
                    return false;
            }
        });
    }

    @Override
    public void startLoading() {
        subscription = Observable.combineLatest(service.schedule(true), service.currentUserIsSignedIn(), toScheduleAndLoggedIn())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(schedule -> updateWith(schedule, this::onEventClicked));
    }

    private BiFunction<Schedule, Boolean, ScheduledAndSignedIn> toScheduleAndLoggedIn() {
        return ScheduledAndSignedIn::new;
    }

    private void onEventClicked(Event event) {
        analytics.trackItemSelected(ContentType.FAVORITES_ITEM, event.id());
        navigate.toEventDetails(event.id());
    }

    @Override
    public void stopLoading() {
        subscription.dispose();
    }

    private void updateWith(ScheduledAndSignedIn scheduledAndSignedIn, ScheduleViewPagerAdapter.OnEventClickedListener listener) {
        if (scheduledAndSignedIn.hasFavorites()) {
            updateWith(scheduledAndSignedIn.schedule, listener);
        } else {
            if (scheduledAndSignedIn.signedIn) {
                promptToFavorite();
            } else {
                promptToSign();
            }
        }
    }

    private void updateWith(Schedule schedule, ScheduleViewPagerAdapter.OnEventClickedListener listener) {
        if (schedule.isEmpty()) {
            promptToFavorite();
        } else {
            favoritesListView.updateWith(schedule, listener);
            favoritesListView.setVisibility(VISIBLE);
            emptyViewSignedIn.setVisibility(GONE);
        }

        progressBar.setVisibility(GONE);
        emptyViewSignedOut.setVisibility(GONE);
    }

    private void promptToFavorite() {
        favoritesListView.setVisibility(GONE);
        progressBar.setVisibility(GONE);
        emptyViewSignedOut.setVisibility(GONE);
        emptyViewSignedIn.setVisibility(VISIBLE);
    }

    public void promptToSign() {
        favoritesListView.setVisibility(GONE);
        progressBar.setVisibility(GONE);
        emptyViewSignedOut.setVisibility(VISIBLE);
        emptyViewSignedIn.setVisibility(GONE);
    }

    private static class ScheduledAndSignedIn {

        final Schedule schedule;
        final boolean signedIn;

        private ScheduledAndSignedIn(Schedule schedule, boolean signedIn) {
            this.schedule = schedule;
            this.signedIn = signedIn;
        }

        boolean hasFavorites() {
            return !schedule.isEmpty();
        }
    }
}

File: app/src/main/java/net/squanchy/favorites/FavoritesPageView.kt
package net.squanchy.favorites

import android.content.Context
import android.util.AttributeSet
import android.view.MenuItem
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.isVisible
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Function3
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.merge_no_favorites_view.view.*
import kotlinx.android.synthetic.main.view_page_favorites.view.*
import net.squanchy.R
import net.squanchy.analytics.Analytics
import net.squanchy.analytics.ContentType
import net.squanchy.favorites.view.FavoritesItem
import net.squanchy.home.HomeActivity
import net.squanchy.home.Loadable
import net.squanchy.navigation.Navigator
import net.squanchy.remoteconfig.FeatureFlags
import net.squanchy.schedule.domain.view.Event
import net.squanchy.support.unwrapToActivityContext
import timber.log.Timber

class FavoritesPageView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : ConstraintLayout(context, attrs, defStyleAttr), Loadable {

    private lateinit var favoritesService: FavoritesService
    private lateinit var navigator: Navigator
    private lateinit var analytics: Analytics
    private lateinit var featureFlags: FeatureFlags

    private val disposable = CompositeDisposable()

    init {
        if (!isInEditMode) {
            with(favoritesComponent(context.unwrapToActivityContext())) {
                favoritesService = favoritesService()
                navigator = navigator()
                analytics = analytics()
                featureFlags = featureFlags()
            }
        }
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        emptyViewSignedOutButton.setOnClickListener { showSignIn() }

        setupToolbar()
    }

    private fun setupToolbar() {
        toolbar.apply {
            title = resources.getString(R.string.activity_favorites)
            inflateMenu(R.menu.homepage)
            setOnMenuItemClickListener(this@FavoritesPageView::onMenuItemClickListener)
        }
    }

    override fun startLoading() {
        disposable.add(
            Observable.combineLatest(
                favoritesService.favorites(),
                favoritesService.currentUserIsSignedIn(),
                featureFlags.showEventRoomInSchedule.toObservable(),
                Function3<List<FavoritesItem>, Boolean, Boolean, LoadResult>(::LoadResult)
            )
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(::handleLoadSchedule, Timber::e)
        )
    }

    override fun stopLoading() = disposable.clear()

    private fun onMenuItemClickListener(menuItem: MenuItem): Boolean {
        return when (menuItem.itemId) {
            R.id.action_search -> {
                showSearch()
                true
            }
            R.id.action_settings -> {
                showSettings()
                true
            }
            else -> false
        }
    }

    private fun handleLoadSchedule(result: LoadResult) {
        when {
            result.favoriteItems.isNotEmpty() -> showFavorites(result.favoriteItems, result.showRoom)
            result.signedIn -> promptToFavorite()
            else -> promptToSign()
        }
    }

    private fun showFavorites(favorites: List<FavoritesItem>, showRoom: Boolean) {
        favoritesListView.updateWith(favorites, showRoom, ::navigateToEventDetails)
        favoritesListView.isVisible = true
        progressBar.isVisible = false
        emptyViewSignedOut.isVisible = false
        emptyViewSignedIn.isVisible = false
    }

    private fun navigateToEventDetails(event: Event) {
        analytics.trackItemSelected(ContentType.FAVORITES_ITEM, event.id)
        navigator.toEventDetails(event.id)
    }

    private fun promptToSign() {
        favoritesListView.isVisible = false
        progressBar.isVisible = false
        emptyViewSignedOut.isVisible = true
        emptyViewSignedIn.isVisible = false
    }

    private fun promptToFavorite() {
        favoritesListView.isVisible = false
        progressBar.isVisible = false
        emptyViewSignedOut.isVisible = false
        emptyViewSignedIn.isVisible = true
    }

    private fun showSignIn() {
        // ⚠️ HACK this is DIRTY and HORRIBLE but it's the only way we can ship this
        // without rewriting the whole data layer. Sorry. I swear, we know it sucks
        // and we want to fix this ASAP.
        val activity: HomeActivity = context.unwrapToActivityContext() as HomeActivity
        activity.requestSignIn()
    }

    private fun showSearch() = navigator.toSearch()

    private fun showSettings() = navigator.toSettings()

    private data class LoadResult(val favoriteItems: List<FavoritesItem>, val signedIn: Boolean, val showRoom: Boolean)
}
